### PR TITLE
Add EDMEulerScheduler

### PR DESCRIPTION
**Pull request by patil-suraj**
# What does this PR do?
This PR is based on #4481 by @dg845 !

This PR adds Euler scheduler with EDM formulation. The difference between this and `EulerDiscreteScheduler` is that

- `EulerDiscreteScheduler` was essentially designed for DDPM models to use euler-style sampling algorithms with discrete timesteps. 
- `EDMEulerScheduler` follows the EDM formulation as closely as possible and is solely intended for models that use EDM formulation, like SVD. It does not support epsilon scaling as that's already covered by `EulerDiscreteScheduler`. So models that still DDPM (the alphas and betas schedules) should keep using `EulerDiscreteScheduler`

**Why add a new scheduler class ?**
While it is possible to support this in  `EulerDiscreteScheduler` by introducing an argument maybe called `scaling_type`, which could correspond to `epsilon (ddpm)` `v-scaling` or `edm`, the issue is that:
- the naming is inconsistent. With `edm`, the model is conditioned on continuous noise scales rather than discrete timesteps. 
- There is some functionality in  `EulerDiscreteScheduler` that is not required for `EDM`, such as computing betas, rescaling for zero terminal SNR, interpolating sigmas etc. 

Hence with the current schedule API if we support full EDM formulation in `EulerDiscreteScheduler`, the code will be confusing to follow with lot's of `if/else` branches. 

**API**

```py
scheduler = EDMEulerScheduler(sigma_min=0.002, sigma_max=80.0, sigma_data=0.5)

scheduler.sigmas # computed directly as per the karras paper 
scheduler.timesteps # this is the `c_noise` which is fed to the model instead of discrete timestep

# to precondition input
sample = scheduler.precondition_inputs(sample, sigma)

# to get the c_noise
c_noise = scheduler.precondition_noise(sigma) 

# precondition output
output = scheduler.precondition_outputs(sample, model_output, sigma)
```

These pre-conditioning methods can be used during training to easily scale the input/output based on sigma values, which the user is free to sample in any way they want during training. 

We could also consider adding a method to help with sigma sampling as per [here](https://github.com/NVlabs/edm/blob/main/training/loss.py#L74)

The rest of the API follows the existing scheduler API. 

I would love to have your thoughts here @yiyixuxu @sayakpaul @dhruvrnaik @dg845

### CODE REVIEW

File: src/diffusers/schedulers/scheduling_edm_euler.py [hunk 1]
```diff
@@ -0,0 +1,412 @@
...
+        Args:
+            model_output (`torch.FloatTensor`):
+                The direct output from learned diffusion model.
+            timestep (`float`):
+                The current discrete timestep in the diffusion chain.
+            sample (`torch.FloatTensor`):
+                A current instance of a sample created by the diffusion process.
+            s_churn (`float`):
+            s_tmin  (`float`):
+            s_tmax  (`float`):
```
Comment (sayakpaul) [inline src/diffusers/schedulers/scheduling_edm_euler.py:307]: Would be nice to have docstrings.

File: src/diffusers/schedulers/scheduling_edm_euler.py [hunk 2]
```diff
@@ -0,0 +1,412 @@
...
+
+        # upon completion increase step index by one
+        self._step_index += 1
+
+        if not return_dict:
+            return (prev_sample,)
+
+        return EDMEulerSchedulerOutput(prev_sample=prev_sample, pred_original_sample=pred_original_sample)
+
+    def add_noise(
```
Comment (sayakpaul) [inline src/diffusers/schedulers/scheduling_edm_euler.py:382]: 👌

File: src/diffusers/schedulers/scheduling_edm_euler.py [hunk 3]
```diff
@@ -0,0 +1,412 @@
...
+    def precondition_outputs(self, sample, model_output, sigma):
+        sigma_data = self.config.sigma_data
+        c_skip = sigma_data**2 / (sigma**2 + sigma_data**2)
+
+        if self.config.prediction_type == "epsilon":
+            c_out = sigma * sigma_data / (sigma**2 + sigma_data**2) ** 0.5
+        elif self.config.prediction_type == "v_prediction":
+            c_out = -sigma * sigma_data / (sigma**2 + sigma_data**2) ** 0.5
+        else:
+            raise ValueError(f"Prediction type {self.config.prediction_type} is not supported.")
```
Comment (sayakpaul) [inline src/diffusers/schedulers/scheduling_edm_euler.py:170]: I guess users will have to do this bit manually during training when experimenting with different prediction types?

**Review by sayakpaul — COMMENTED** [review]
Design looks very nice to me!

**Review by yiyixuxu — APPROVED** [review]
looking good to me:) 

question: 
for `v_prediction`: does it work the same as in what's currently implemented in `EulerDiscreteScheduler`?

File: src/diffusers/schedulers/scheduling_edm_euler.py [hunk 4]
```diff
@@ -0,0 +1,412 @@
...
+    def __init__(
+        self,
+        sigma_min: float = 0.002,
+        sigma_max: float = 80.0,
+        sigma_data: float = 0.5,
+        num_train_timesteps: int = 1000,
+        prediction_type: str = "epsilon",
+        rho: float = 7.0,
+        timestep_spacing: str = "linspace",
+        steps_offset: int = 0,
```
Comment (yiyixuxu) [inline src/diffusers/schedulers/scheduling_edm_euler.py:98]:
```suggestion
```
do we need them?

File: src/diffusers/schedulers/scheduling_edm_euler.py (hunk 2 above)
Comment (yiyixuxu) [inline src/diffusers/schedulers/scheduling_edm_euler.py:382]: add a `#Copied from` I guess

File: src/diffusers/schedulers/scheduling_edm_euler.py [hunk 5]
```diff
@@ -0,0 +1,412 @@
...
+        elif self.config.timestep_spacing == "trailing":
+            step_ratio = self.config.num_train_timesteps / self.num_inference_steps
+            # creates integer timesteps by multiplying by ratio
+            # casting to int to avoid issues when num_inference_step is power of 3
+            timesteps = (np.arange(self.config.num_train_timesteps, 0, -step_ratio)).round().copy().astype(np.float32)
+            timesteps -= 1
+        else:
+            raise ValueError(
+                f"{self.config.timestep_spacing} is not supported. Please make sure to choose one of 'linspace', 'leading' or 'trailing'."
+            )
```
[the remark this answers is not in the document]
Comment (patil-suraj) [reply src/diffusers/schedulers/scheduling_edm_euler.py:234]: Good catch, not really required.

File: src/diffusers/schedulers/scheduling_edm_euler.py (hunk 4 above)
Comment (patil-suraj) [reply src/diffusers/schedulers/scheduling_edm_euler.py:98]: Nope

**Comment by patil-suraj** [comment]
> for v_prediction: does it work the same as in what's currently implemented in EulerDiscreteScheduler?

@yiyixuxu  Yes, that's right. `EulerDiscreteScheduler` would work the same way if `sigma_min` , `sigma_max` is provided and `timestep_type==continuous`


Actually not sure if we should put it here, it's used to SVD and follows EDM as well.

**Comment by sayakpaul** [comment]
I think okay to put it there as it helps the community to experiment with different prediction types and provide us feedback if needed. We can iterate on top of that.

**Pull request merged**